### ISSUE TITLE
[BUG] Exception specifications on fused functions are too general

### ISSUE THREAD

**Issue by da-woods** [cython/cython#5586]
### Describe the bug

Cython seems to use `except *` for fused functions rather than the more sensible `except -1` (or similar) that the specialized type would imply.

### Code to reproduce the behaviour:

```cython
cimport cython

cpdef cython.numeric f(cython.numeric x):
    return x

def test_f_double(double x):
    return f(x)

def test_f_int(int x):
    return f(x)
```

The generated C code is:

```c
__pyx_t_1 = __pyx_fuse_4__pyx_f_13fused_rettype_f(__pyx_v_x, 0); if (unlikely(PyErr_Occurred())) __PYX_ERR(0, 7, __pyx_L1_error)
```

and 

```c
__pyx_t_1 = __pyx_fuse_1__pyx_f_13fused_rettype_f(__pyx_v_x, 0); if (unlikely(PyErr_Occurred())) __PYX_ERR(0, 10, __pyx_L1_error)
```

i.e. it misses the cheap `if (__pyx_t_1 == -1)` call that it should be inserting.

### Expected behaviour

The fused functions should pick up the default error return value of their specializations.

### OS

Linux

### Python version

3.11

### Cython version

3.0.0

### Additional context

Related to https://github.com/cython/cython/issues/5564 - that issue specifically involves Scipy fused functions that are a lot more costly in a `nogil` block than they should be because of this (but that issue is mainly about a deadlock).

I'd like to get this into 3.0.1 if possible.

**Comment by da-woods** [comment]
This is currently looking a bit fiddly...

**Comment by da-woods** [comment]
There may be a problem with this: https://github.com/cython/cython/pull/5593#issuecomment-1669391312 - it's a breaking change since in Cython 3.0.0 the functions would return `0` rather than `-1`.

Therefore I think it can only safely be applied to functions that aren't defined in a separate pxd file (which doesn't really help the scipy case). Probably still worth doing, but less benefit I think.

**Comment by da-woods** [comment]
Potentially we could make the test for externally defined functions `result == 0 || result == -1` which would likely be better than nothing, but doesn't fit well into the existing system

**Comment by matusvalo** [comment]
> it's a breaking change since in Cython 3.0.0 the functions would return 0 rather than -1.

I am not sure whether I understand correctly. Where the funcgtions are returning -1? In Cython 3.0 are returning 0:

<img width="715" alt="image" src="https://github.com/cython/cython/assets/827060/71dbddba-b853-4767-a56f-e5066f37c757">

Is there any PR changing that? Maybe, I have missed something...

**Comment by da-woods** [comment]
You're right that they're running `0` (since that's the "default value" for the numeric types).

However, typically we'd use `-1` as an error code for numeric types 

My proposed change here was to make fused functions use the usual `-1` error code. (A PR doesn't yet exist). The problem would be that you couldn't assume `-1` for functions that came from a pxd file because they might have been generated with an earlier version of Cython.

**Comment by matusvalo** [comment]
> You're right that they're running 0 (since that's the "default value" for the numeric types).
>
> However, typically we'd use -1 as an error code for numeric types

Yeah, now I see:
![image](https://github.com/cython/cython/assets/827060/715f6b71-4fb7-40c6-a6f2-3ef2f3752226)

Functions returning numbers are propagating exception by default via `-1` and pointers/python objects propagate exceptions by default via `0`.

A question, is there any complication simple to propagate for fused types via `0`? It's not 100% consistent but i consider it still more straightforward than having special logic for `.pxd` files or having two values for propagation (`0` and `-1`)?

**Comment by da-woods** [comment]
> A question, is there any complication simple to propagate for fused types via `0`? It's not 100% consistent but i consider it still more straightforward than having special logic for `.pxd` files or having two values for propagation (`0` and `-1`)?

No - the options that I'm considering are:

* Use `0` as the error value for fused functions (as you suggest). The disadvantages are

  * it's inconsistent (so it'll catch a few people out making function pointers etc)
  * `0` is probably a more common return value than `-1` so it'll probably perform work (but still better than `except *`).
    
  The advantages are that it works now.

* Use `0` as the value for fused functions defined in pxd files and `-1` elsewhere. The disadvantages are

  * it's *very* inconsistent
  * Probably quite complicated to implement

* Return `-1` from all fused functions. Assume `except *` when calling the function if there isn't a local definition. At some point in the future when nobody's using Cython 3.0.0 we can switch the check everywhere.

  * The advantage is that it ends up consistent eventually.
  * The disadvantage is that it doesn't perform well now.

* Return `-1` from all fused functions. Check for `-1` *and* `0` for fused functions where there isn't a local definition. Switch at some point in the future.
  * Advantage: ends up consistent, performs OK now
  * Disadvantage: hard to implement.

I'm really not sure what the best option is (except I really don't like the second one). Still thinking. I have about 50% of an implementation, but haven't had time to finish it yet
